### ISSUE TITLE
BinaryBuilder script for HDF5

### ISSUE THREAD

**Issue by musm** [JuliaIO/HDF5.jl#471]
Would be nice to update to using a binarybuilder script to provide a consistent hdf5 version across all platforms

**Comment by stevengj** [comment]
I put together a first draft here: https://github.com/stevengj/HDF5Builder

**Comment by stevengj** [comment]
Should also help with #485 because we can build libhdf5 to use its own internal zlib rather than the system zlib.

**Comment by stevengj** [comment]
Unfortunately, it looks like the HDF5 build script makes liberal use of `TRY_RUN` to detect binary formats.  It is possible to override, but we'll need to figure out how to set a bunch of low-level cache variables:
```
CMake Error: TRY_RUN() invoked in cross-compiling mode, please set the following cache variables appropriately:
   H5_PRINTF_LL_TEST_RUN (advanced)
   H5_PRINTF_LL_TEST_RUN__TRYRUN_OUTPUT (advanced)
For details see /workspace/srcdir/CMake-hdf5-1.8.20/build/TryRunResults.cmake
Width test failed with result: PLEASE_FILL_OUT-FAILED_TO_RUN
-- Checking for appropriate format for 64 bit long: not found
-- Looking for difftime
-- Looking for difftime - found
-- Looking for clock_gettime
-- Looking for clock_gettime - not found
-- Looking for clock_gettime in rt
-- Looking for clock_gettime in rt - found
-- Looking for clock_gettime in posix4
-- Looking for clock_gettime in posix4 - not found
CMake Error: TRY_RUN() invoked in cross-compiling mode, please set the following cache variables appropriately:
   H5_LDOUBLE_TO_LONG_SPECIAL_RUN (advanced)
   H5_LDOUBLE_TO_LONG_SPECIAL_RUN__TRYRUN_OUTPUT (advanced)
For details see /workspace/srcdir/CMake-hdf5-1.8.20/build/TryRunResults.cmake
-- Checking IF your system converts long double to (unsigned) long values with special algorithm... no
CMake Error: TRY_RUN() invoked in cross-compiling mode, please set the following cache variables appropriately:
   H5_LONG_TO_LDOUBLE_SPECIAL_RUN (advanced)
   H5_LONG_TO_LDOUBLE_SPECIAL_RUN__TRYRUN_OUTPUT (advanced)
For details see /workspace/srcdir/CMake-hdf5-1.8.20/build/TryRunResults.cmake
-- Checking IF your system can convert (unsigned) long to long double values with special algorithm... no
CMake Error: TRY_RUN() invoked in cross-compiling mode, please set the following cache variables appropriately:
   H5_LDOUBLE_TO_LLONG_ACCURATE_RUN (advanced)
   H5_LDOUBLE_TO_LLONG_ACCURATE_RUN__TRYRUN_OUTPUT (advanced)
For details see /workspace/srcdir/CMake-hdf5-1.8.20/build/TryRunResults.cmake
-- Checking IF correctly converting long double to (unsigned) long long values... no
CMake Error: TRY_RUN() invoked in cross-compiling mode, please set the following cache variables appropriately:
   H5_LLONG_TO_LDOUBLE_CORRECT_RUN (advanced)
   H5_LLONG_TO_LDOUBLE_CORRECT_RUN__TRYRUN_OUTPUT (advanced)
For details see /workspace/srcdir/CMake-hdf5-1.8.20/build/TryRunResults.cmake
-- Checking IF correctly converting (unsigned) long long to long double values... no
CMake Error: TRY_RUN() invoked in cross-compiling mode, please set the following cache variables appropriately:
   H5_NO_ALIGNMENT_RESTRICTIONS_RUN (advanced)
   H5_NO_ALIGNMENT_RESTRICTIONS_RUN__TRYRUN_OUTPUT (advanced)
For details see /workspace/srcdir/CMake-hdf5-1.8.20/build/TryRunResults.cmake
-- Checking IF alignment restrictions are strictly enforced... no
-- Filter ZLIB is built
-- Filter ZLIB is ON
-- Filter SZIP is built
-- Filter SZIP is ON
-- Configuring incomplete, errors occurred!
See also "/workspace/srcdir/CMake-hdf5-1.8.20/build/CMakeFiles/CMakeOutput.log".
See also "/workspace/srcdir/CMake-hdf5-1.8.20/build/CMakeFiles/CMakeError.log".
ERROR: LoadError: Build for HDF5 on i686-linux-gnu did not complete successfully
```

**Comment by ararslan** [comment]
@stevengj, looks like we made HDF5 builders at almost exactly the same time: https://github.com/ararslan/HDF5Builder. I noticed that the HDF Group [explicitly states](https://portal.hdfgroup.org/pages/viewpage.action?pageId=48808266) that HDF5 does not support cross compilation, which makes using BinaryBuilder for this a bit questionable.

**Comment by stevengj** [comment]
It's too bad, especially since it looks like almost all of the things that they use `TRY_RUN` for are things that can be detected statically by examining compiler output (e.g. [endian-ness](https://www.gnu.org/software/autoconf-archive/ax_c_float_words_bigendian.html)).

**Comment by ararslan** [comment]
BinaryBuilder supports using custom patches now IIRC, so if we know what needs to change in the source itself in order to cross compile HDF5, we could potentially maintain a set of patches to support it until the changes upstreamed and released.

There actually is a public Git repository for HDF5 on Bitbucket: https://bitbucket.hdfgroup.org/projects/HDFFV/repos/hdf5/browse.

**Comment by musm** [comment]
I was really looking forward too this, too bad HDF5 does not support cross compilation.

**Comment by musm** [comment]
So on Julia 1.0, this package can no longer be installed on Windows and there is a long standing issue on Mac that seems to prevent a large number of people using this package. 

How realistic is it to just upload the binaries from MacOs and Windows from https://www.hdfgroup.org/downloads/hdf5/?

**Comment by ggggggggg** [comment]
Could we just have people with existing installations package up the needed files and attach them on this issue? I have a working OSX installation, I'm not actually sure the exact files that are needed, but I'm happy to upload some files with a bit of guidance. For now I've zipped up everything that started with "libhdf5" in `/Users/user/.julia/v0.6/Homebrew/deps/usr/lib/`.
[libhdf5.zip](https://github.com/JuliaIO/HDF5.jl/files/2303089/libhdf5.zip)

**Comment by musm** [comment]
@ggggggggg  If you download the  pre-built binaries from https://www.hdfgroup.org/downloads/hdf5/ 

and manually modify the data/deps.jl file  by changing Line 20. Does the test-suite pass?

For example I downloaded the pre-built binaries and manually modified the deps.jl to point to it.
`@checked_lib libhdf5 "C:\\Program Files\\HDF_Group\\HDF5\\1.10.2\\bin\\hdf5.dll"` 
There are segfaults on Line83 in plain.jl due to ` g["BloscA", "chunk", (5, 6), "shuffle", (), "blosc", 9] = R` and some other issues, but it runs fine.

Why can't we just use BinaryProvider to provide pre-built packages. It won't be automated a la BinaryBuilder, but  that's fine. We can always build manually and upload.

**Comment by ggggggggg** [comment]
I'm still on 0.6, I won't have time to upgrade or test things soon.

**Comment by musm** [comment]
There is the following patch list if someone would like to try (I won't be able to in the immediate future, maybe sometime  later)

https://github.com/Alexpux/MINGW-packages/tree/master/mingw-w64-hdf5

**Comment by stevengj** [comment]
I haven't had a chance to work on this for a while, but the current state of my effort at getting HDF5 to cross-compile properly with its `configure` script is here:   https://github.com/stevengj/hdf5

I'm working my slowly through `configure.ac` and so far haven't encountered any insurmountable obstacles — it has been possible to rewrite every problematic test in terms of Autoconf's cross-compile-safe feature tests.

**Comment by visr** [comment]
@stevengj have you seen this? https://github.com/live-clones/hdf5/blob/a9283319a942032022384a1c3560a3f92ded36bf/release_docs/RELEASE.txt#L60-L67

The release notes for the HDF5 version under development lists "Add toolchain and cross-compile support" under new features.

**Comment by musm** [comment]
https://portal.hdfgroup.org/display/support/HDF5%201.10.4#releasenotes

**Comment by musm** [comment]
Seems that cross-compile support has been added with version 1.10.4

**Comment by omus** [comment]
Issue should be closed due to #555
